### PR TITLE
Dynamic Module

### PR DESCRIPTION
**Pull request by MarcusTXK**
### Description

Updated the current SideContentHelper.ts to utilise the [new modules system](https://github.com/source-academy/modules/tree/modules/development)

 In `SideContentHelper`, `getDynamicTabs` recieves the debuggerContext and extracts the modules from it (using  `getModuleTabs`), and filters the modules based on their toSpawn condition. Side content tabs will only spawn if the `toSpawn` condition evaluates to true.

The relevant types needed were also added in `src/commons/sideContent/SideContentTypes.ts`.

### Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update
- [ ] Code quality improvements

### How to test
1) Pull js-slang from modules/development branch, and [follow the steps here](https://github.com/source-academy/js-slang#using-your-js-slang-in-local-source-academy) to link it to the cadet-front end

2) Pull cadet-frontend from modules/development branch, and setup the env file and run it. Set the module backend url to this:
`REACT_APP_MODULE_BACKEND_URL=http://localhost:8022`

3) Pull modules from modules/development branch, and [follow the steps here](https://github.com/source-academy/modules/tree/modules/development) to set it up and run it.

4) On the playground, the side content tab will spawn after the 'test' module is imported and the output in the REPL is "test":
image

### Checklist
- [x] I have tested this code

### CODE REVIEW

**Comment by MarcusTXK** [comment]
`Run CI / Run CI (pull_request) Failing after 1m`
This is occurring as there are changes made in both js-slang and cadet-frontend. 
`modules` has been added to the `Context` object in js-slang (at the bottom)
image

**Review by anthony-halim — COMMENTED** [review]
Will there be plans to remove SideContentTab in the future? It appears ModuleSideContent is taking up the same role.

File: src/commons/sideContent/SideContentTypes.ts [hunk 1]
```diff
@@ -56,3 +57,19 @@ export type SideContentTab = {
   id?: SideContentType;
   disabled?: boolean;
 };
+
```
Comment (anthony-halim) [inline src/commons/sideContent/SideContentTypes.ts:60]: Documentation needed

File: src/commons/sideContent/SideContentTypes.ts [hunk 2]
```diff
@@ -56,3 +57,19 @@ export type SideContentTab = {
...
+  label: string;
+  iconName: IconName;
+  /**
+   * @todo Convert the any type for React parameter to React type
+   * @todo Convert the any type for props parameter
+   */
+  body: (React: any) => (props: any) => JSX.Element;
+  toSpawn: (context: DebuggerContext) => boolean;
+};
+
```
Comment (anthony-halim) [inline src/commons/sideContent/SideContentTypes.ts:71]: Documentation needed

**Comment by MarcusTXK** [comment]
> Will there be plans to remove SideContentTab in the future? It appears ModuleSideContent is taking up the same role.

Apologies for the late reply, currently there are no plans to merge them as modules imported from the modules repo will have a slightly different structure than the current SideContentTab, eg. for the body, React needs to be passed into the body for content tabs imported modules

**Comment by anthony-halim** [comment]
> > Will there be plans to remove SideContentTab in the future? It appears ModuleSideContent is taking up the same role.
> 
> Apologies for the late reply, currently there are no plans to merge them as modules imported from the modules repo will have a slightly different structure than the current SideContentTab, eg. for the body, React needs to be passed into the body for content tabs imported modules

SideContentTab is essentially an empty system (no module tab has migrated into it); would it be possible for ModuleSideContent to accomodate the old system and replace SideContentTab? Having two systems doing the same purpose will only clutter the codebase.

**Comment by MarcusTXK** [comment]
> > > Will there be plans to remove SideContentTab in the future? It appears ModuleSideContent is taking up the same role.
> > 
> > 
> > Apologies for the late reply, currently there are no plans to merge them as modules imported from the modules repo will have a slightly different structure than the current SideContentTab, eg. for the body, React needs to be passed into the body for content tabs imported modules
> 
> SideContentTab is essentially an empty system (no module tab has migrated into it); would it be possible for ModuleSideContent to accomodate the old system and replace SideContentTab? Having two systems doing the same purpose will only clutter the codebase.

Alright sure. We are still tweaking the ModuleSideContent type currently to find the best way to accommodate the features of the dynamic modules so we will temporarily leave it there first. Once we are settled on the ModuleSideContents then we will deprecate the old system because there are still limitations to our new system.

**Review by anthony-halim — APPROVED** [review]
LGTM!

**Pull request merged**